x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/WatcherLifeCycleService.java

```diff
@@ -99,8 +99,15 @@ public void clusterChanged(ClusterChangedEvent event) {
         if (isWatcherStoppedManually) {
             if (this.state.get() == WatcherState.STARTED) {
                 clearAllocationIds();
-                watcherService.stop("watcher manually marked to shutdown by cluster state update");
-                this.state.set(WatcherState.STOPPED);
+                this.state.set(WatcherState.STOPPING);
+                //waiting to set state to stopped until after all currently running watches are finished
+                watcherService.stop("watcher manually marked to shutdown by cluster state update", a -> {
+                    //ensure that Watcher wasn't restarted between stopping and now.
+                    if (state.get() == WatcherState.STOPPING) {
```

Comment:
this looks racy to me. As it is an atomicreference you can set it atomically and may want to log something if the state was not set properly?

Reply:
agreed, a `compareAndSet(...)` is preferable here. Also I think `compareAndSet(...)` can be used to go from started to stopping a few lines above here?

Reply:
done in https://github.com/elastic/elasticsearch/pull/43939/commits/39b0336b38555d3a2590c063068cd595ee327210